File: app/src/main/java/app/trigger/mqtt/MqttRequestHandler.java
package app.trigger.mqtt;

import javax.net.ssl.SSLContext;
import javax.net.ssl.X509TrustManager;
import javax.security.cert.CertificateException;

import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
import org.eclipse.paho.client.mqttv3.MqttCallback;
import org.eclipse.paho.client.mqttv3.MqttClient;
import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
import org.eclipse.paho.client.mqttv3.MqttException;
import org.eclipse.paho.client.mqttv3.MqttMessage;
import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence;

import java.security.PrivateKey;
import java.security.SecureRandom;
import java.security.cert.X509Certificate;

import app.trigger.Utils;
import app.trigger.MainActivity.Action;
import app.trigger.MqttDoorSetup;
import app.trigger.DoorReply;
import app.trigger.DoorReply.ReplyCode;
import app.trigger.OnTaskCompleted;
import app.trigger.Log;
import app.trigger.WifiTools;
import app.trigger.https.HttpsTools;
import app.trigger.ssh.PubkeyUtils;


public class MqttRequestHandler extends Thread implements MqttCallback {
    private static final String TAG = "MqttRequestHandler";
    private final OnTaskCompleted listener;
    private final MqttDoorSetup setup;
    private final Action action;

    public MqttRequestHandler(OnTaskCompleted listener, MqttDoorSetup setup, Action action) {
        this.listener = listener;
        this.setup = setup;
        this.action = action;
    }

    public void run() {
        if (setup.getId() < 0) {
            this.listener.onTaskResult(setup.getId(), ReplyCode.LOCAL_ERROR, "Internal Error");
            return;
        }

        if (Utils.isEmpty(setup.server)) {
            this.listener.onTaskResult(setup.getId(), ReplyCode.LOCAL_ERROR, "MQTT broker address not set.");
            return;
        }

        if (action == Action.fetch_state && setup.status_topic.isEmpty()) {
            this.listener.onTaskResult(setup.getId(), ReplyCode.LOCAL_ERROR, "");
            return;
        }

        if ((action == Action.open_door || action == Action.ring_door || action == Action.close_door)
                && setup.command_topic.isEmpty()) {
            this.listener.onTaskResult(setup.getId(), ReplyCode.LOCAL_ERROR, "No command topic set.");
            return;
        }

        if (action == Action.close_door && setup.close_command.isEmpty()) {
            this.listener.onTaskResult(setup.getId(), ReplyCode.LOCAL_ERROR, "No close command set.");
            return;
        }

        if (setup.qos != 0 && setup.qos != 1 && setup.qos != 2) {
            this.listener.onTaskResult(setup.getId(), ReplyCode.LOCAL_ERROR, "Invalid QoS value: " + setup.qos);
            return;
        }

        String clientId = MqttClient.generateClientId();
        MqttConnectOptions opts = new MqttConnectOptions();
        MemoryPersistence persistence = new MemoryPersistence();
        String address = setup.server;

        try {
            if (address.startsWith("mqtt://")) {
                address = address.replaceFirst("mqtt://", "tcp://");
                address = Utils.rebuildAddress(address, 1883);
            } else if (address.startsWith("mqtts://")) {
                address = address.replaceFirst("mqtts://", "ssl://");
                address = Utils.rebuildAddress(address, 8883);
            } else if (setup.server.startsWith("tcp://")) {
                address = Utils.rebuildAddress(address, 1883);
            } else if (setup.server.startsWith("ssl://")) {
                address = Utils.rebuildAddress(address, 8883);
            } else {
                throw new Exception(
                    "Server address needs to start with 'mqtt://' or 'mqtts://'."
                );
            }

            if (!setup.username.isEmpty()) {
                opts.setUserName(setup.username);
            }

            if (!setup.password.isEmpty()) {
                opts.setPassword(setup.password.toCharArray());
            }

            // false: broker will not keep messages
            opts.setCleanSession(false);

            if (address.startsWith("ssl://")) {
                opts.setHttpsHostnameVerificationEnabled(!setup.ignore_hostname_mismatch);

                if (setup.server_certificate != null) {
                    // use given certificate only
                    if (setup.client_keypair != null && setup.client_certificate != null) {
                        PrivateKey client_private_key = PubkeyUtils.decodePrivate(
                            setup.client_keypair.getPrivateKey(), setup.client_keypair.getType()
                        );
                        opts.setSocketFactory(
                            Utils.getSocketFactoryWithCertificateAndClientKey(
                                setup.server_certificate, setup.client_certificate, client_private_key)
                        );
                    } else if (setup.client_keypair == null && setup.client_certificate == null) {
                        if (setup.ignore_certificate) {
                            // disable entire certificate validity
                            SSLContext context = SSLContext.getInstance("TLS");
                            context.init(null, new X509TrustManager[]{new X509TrustManager() {
                                @Override
                                public void checkClientTrusted(java.security.cert.X509Certificate[] chain, String authType) {}

                                @Override
                                public void checkServerTrusted(java.security.cert.X509Certificate[] chain, String authType) {}

                                @Override
                                public X509Certificate[] getAcceptedIssuers() {
                                    return new X509Certificate[0];
                                }}}, new SecureRandom());
                            opts.setSocketFactory(context.getSocketFactory());
                        } else if (setup.ignore_expiration) {
                            // ignore notBefore/notAfter
                            opts.setSocketFactory(
                                    HttpsTools.getSocketFactoryIgnoreCertificateExpiredException()
                            );
                        } else {
                            opts.setSocketFactory(
                                Utils.getSocketFactoryWithCertificate(setup.server_certificate)
                            );
                        }
                    } else {
                        throw new Exception("Both client key and client certificate needed.");
                    }
                } else {
                    if (setup.client_keypair != null || setup.client_certificate != null) {
                        throw new Exception("Client key and client certificate needed.");
                    } else {
                        // use system default certificates
                        opts.setSocketFactory(
                            SSLContext.getDefault().getSocketFactory()
                        );
                    }
                }
            }

            MqttClient client = new MqttClient(address, clientId, persistence);
            client.setTimeToWait(3000); // 3 seconds
            client.setCallback(this);
            client.connect(opts);

            switch (action) {
                case fetch_state:
                    // subscribe
                    client.subscribe(setup.status_topic);
                    break;
                case open_door:
                    // publish
                    MqttMessage open = new MqttMessage(setup.open_command.getBytes());
                    open.setRetained(setup.retained);
                    open.setQos(setup.qos);

                    client.publish(setup.command_topic, open);
                    break;
                case ring_door:
                    // publish
                    MqttMessage ring = new MqttMessage(setup.ring_command.getBytes());
                    ring.setRetained(setup.retained);
                    ring.setQos(setup.qos);

                    client.publish(setup.command_topic, ring);
                    break;
                case close_door:
                    // publish
                    MqttMessage close = new MqttMessage(setup.close_command.getBytes());
                    close.setRetained(setup.retained);
                    close.setQos(setup.qos);

                    client.publish(setup.command_topic, close);
                    break;
            }
            this.listener.onTaskResult(setup.getId(), ReplyCode.SUCCESS, "");
        } catch (MqttException me) {
            //me.getMessage() returns "MqttException" only
            this.listener.onTaskResult(setup.getId(), ReplyCode.REMOTE_ERROR, me.toString());
        } catch (Exception e) {
            this.listener.onTaskResult(setup.getId(), ReplyCode.LOCAL_ERROR, e.getMessage());
        }
    }

    @Override
    public void connectionLost(Throwable cause) {
        listener.onTaskResult(setup.getId(), ReplyCode.REMOTE_ERROR, cause.toString());
    }

    @Override
    public void messageArrived(String topic, final MqttMessage message) {
        listener.onTaskResult(setup.getId(), ReplyCode.SUCCESS, message.toString());
    }

    @Override
    public void deliveryComplete(IMqttDeliveryToken token) {
        // ignore
    }

    protected void onPostExecute(DoorReply result) {
        // already on GUI thread
        listener.onTaskResult(setup.getId(), result.code, result.message);
    }
}

File: app/src/main/kotlin/app/trigger/mqtt/MqttRequestHandler.kt
package app.trigger.mqtt

import app.trigger.Utils.getSocketFactoryWithCertificateAndClientKey
import app.trigger.Utils.getSocketFactoryWithCertificate
import app.trigger.Utils.rebuildAddress
import app.trigger.DoorReply.ReplyCode
import app.trigger.https.HttpsTools
import app.trigger.ssh.PubkeyUtils
import org.eclipse.paho.client.mqttv3.MqttCallback
import org.eclipse.paho.client.mqttv3.MqttClient
import org.eclipse.paho.client.mqttv3.MqttConnectOptions
import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence
import org.eclipse.paho.client.mqttv3.MqttMessage
import org.eclipse.paho.client.mqttv3.MqttException
import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken
import app.trigger.*
import java.lang.Exception
import java.security.SecureRandom
import java.security.cert.X509Certificate
import javax.net.ssl.SSLContext
import javax.net.ssl.X509TrustManager

class MqttRequestHandler(private val listener: OnTaskCompleted, private val door: MqttDoor, private val action: MainActivity.Action)
        : Thread(), MqttCallback {
    override fun run() {
        if (door.id < 0) {
            listener.onTaskResult(door.id, action, ReplyCode.LOCAL_ERROR, "Internal Error")
            return
        }
        if (door.server.isEmpty()) {
            listener.onTaskResult(door.id, action, ReplyCode.LOCAL_ERROR, "MQTT broker address not set.")
            return
        }
        if (action === MainActivity.Action.FETCH_STATE && door.status_topic.isEmpty()) {
            listener.onTaskResult(door.id, action, ReplyCode.LOCAL_ERROR, "")
            return
        }
        if ((action === MainActivity.Action.OPEN_DOOR || action === MainActivity.Action.RING_DOOR || action === MainActivity.Action.CLOSE_DOOR)
                && door.command_topic.isEmpty()) {
            listener.onTaskResult(door.id, action, ReplyCode.LOCAL_ERROR, "No command topic set.")
            return
        }
        if (action === MainActivity.Action.CLOSE_DOOR && door.close_command.isEmpty()) {
            listener.onTaskResult(door.id, action, ReplyCode.LOCAL_ERROR, "No close command set.")
            return
        }
        if (door.qos != 0 && door.qos != 1 && door.qos != 2) {
            listener.onTaskResult(door.id, action, ReplyCode.LOCAL_ERROR, "Invalid QoS value: ${door.qos}")
            return
        }
        val clientId = MqttClient.generateClientId()
        val opts = MqttConnectOptions()
        val persistence = MemoryPersistence()
        var address = door.server
        try {
            if (address.startsWith("mqtt://")) {
                address = address.replaceFirst("mqtt://".toRegex(), "tcp://")
                address = rebuildAddress(address, 1883)
            } else if (address.startsWith("mqtts://")) {
                address = address.replaceFirst("mqtts://".toRegex(), "ssl://")
                address = rebuildAddress(address, 8883)
            } else if (door.server.startsWith("tcp://")) {
                address = rebuildAddress(address, 1883)
            } else if (door.server.startsWith("ssl://")) {
                address = rebuildAddress(address, 8883)
            } else {
                throw Exception(
                        "Server address needs to start with 'mqtt://' or 'mqtts://'."
                )
            }
            if (door.username.isNotEmpty()) {
                opts.userName = door.username
            }
            if (door.password.isNotEmpty()) {
                opts.password = door.password.toCharArray()
            }

            // false: broker will not keep messages
            opts.isCleanSession = false
            if (address.startsWith("ssl://")) {
                opts.isHttpsHostnameVerificationEnabled = !door.ignore_hostname_mismatch
                val client_keypair = door.client_keypair
                val client_certificate = door.client_certificate
                if (door.server_certificate != null) {
                    // use given certificate only
                    if (client_keypair != null && client_certificate != null) {
                        val client_private_key = PubkeyUtils.decodePrivate(
                                client_keypair.privateKey, client_keypair.type
                        )
                        opts.socketFactory = getSocketFactoryWithCertificateAndClientKey(
                                door.server_certificate, door.client_certificate!!, client_private_key)
                    } else if (client_keypair == null && client_certificate == null) {
                        if (door.ignore_certificate) {
                            // disable entire certificate validity
                            val context = SSLContext.getInstance("TLS")
                            context.init(null, arrayOf<X509TrustManager>(object : X509TrustManager {
                                override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) {}
                                override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) {}
                                override fun getAcceptedIssuers(): Array<X509Certificate> {
                                    return arrayOf<X509Certificate>()
                                }
                            }), SecureRandom())
                            opts.socketFactory = context.socketFactory
                        } else if (door.ignore_expiration) {
                            // ignore notBefore/notAfter
                            opts.socketFactory = HttpsTools.getSocketFactoryIgnoreCertificateExpiredException()
                        } else {
                            opts.socketFactory = getSocketFactoryWithCertificate(door.server_certificate)
                        }
                    } else {
                        throw Exception("Both client key and client certificate needed.")
                    }
                } else {
                    if (client_keypair != null || client_certificate != null) {
                        throw Exception("Client key and client certificate needed.")
                    } else {
                        // use system default certificates
                        opts.socketFactory = SSLContext.getDefault().socketFactory
                    }
                }
            }
            val client = MqttClient(address, clientId, persistence)
            client.timeToWait = 3000 // 3 seconds
            client.setCallback(this)
            client.connect(opts)
            when (action) {
                MainActivity.Action.FETCH_STATE ->                     // subscribe
                    client.subscribe(door.status_topic)
                MainActivity.Action.OPEN_DOOR -> {
                    // publish
                    val message = MqttMessage(door.open_command.toByteArray())
                    message.isRetained = door.retained
                    message.qos = door.qos
                    client.publish(door.command_topic, message)
                }
                MainActivity.Action.RING_DOOR -> {
                    // publish
                    val message = MqttMessage(door.ring_command.toByteArray())
                    message.isRetained = door.retained
                    message.qos = door.qos
                    client.publish(door.command_topic, message)
                }
                MainActivity.Action.CLOSE_DOOR -> {
                    // publish
                    val message = MqttMessage(door.close_command.toByteArray())
                    message.isRetained = door.retained
                    message.qos = door.qos
                    client.publish(door.command_topic, message)
                }
            }
            listener.onTaskResult(door.id, action, ReplyCode.SUCCESS, "")
        } catch (me: MqttException) {
            //me.getMessage() returns "MqttException" only
            listener.onTaskResult(door.id, action, ReplyCode.REMOTE_ERROR, me.toString())
        } catch (e: Exception) {
            listener.onTaskResult(door.id, action, ReplyCode.LOCAL_ERROR, e.message!!)
        }
    }

    override fun connectionLost(cause: Throwable) {
        listener.onTaskResult(door.id, action, ReplyCode.REMOTE_ERROR, cause.toString())
    }

    override fun messageArrived(topic: String, message: MqttMessage) {
        listener.onTaskResult(door.id, action, ReplyCode.SUCCESS, message.toString())
    }

    override fun deliveryComplete(token: IMqttDeliveryToken) {
        // ignore
    }

    protected fun onPostExecute(result: DoorReply) {
        // already on GUI thread
        listener.onTaskResult(door.id, action, result.code, result.message)
    }

    companion object {
        private const val TAG = "MqttRequestHandler"
    }
}